NIFI-12898 Allow uploading asset and referencing via Parameter

bbende opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

This pull request adds the ability to upload assets to parameter contexts and then reference them from parameters in the context. This feature is described in [NIFI-12898](https://issues.apache.org/jira/browse/NIFI-12898) and the [feature proposal](https://cwiki.apache.org/confluence/display/NIFI/Asset+Management).

A new framework extension - `AssetManager` -  is responsible for storing/retrieving assets on the file system. 

The following REST end-points are added:

Create/upload an asset:
`POST /parameter-contexts/{id}/assets`  

List assets:
`GET  /parameter-contexts/{id}/assets`

Get asset content:
`GET  /parameter-contexts/{id}/assets/{assetId}`

The last two end-points are primarily intended to be used when a node joins the cluster for it to compare it's current assets with the coordinator and synchronize it's assets to match.

To reference an asset it is done through a normal parameter context update, where a parameter will send no value and instead provide an asset reference:
```
referencedAssets: [
  {
     "id": "assetId1"
  },
 {
     "id": "assetId2"
  },
]
```

The back-end will automatically make the value of the parameter become the concatenated paths of the referenced assets.

The following CLI commands were added:

```
nifi create-asset -pcid <param-context-id> -af <full-path-to-a-file>

nifi list-assets -pcid <param-context-id>

nifi add-asset-reference -pcid <param-context-id> -pn <param-name> -aid <asset-id>

nifi remove-asset-reference -pcid <param-context-id> -pn <param-name> -aid <asset-id>
```

When a parameter context is updated, any asset that was previously referenced by a parameter and no longer referenced after the update will be deleted.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

exceptionfactory reviewed (commented):
Thanks for putting together this new feature @bbende! I plan to take a closer look soon. 

GitHub CodeQL flagged several issues related to using input parameters for file names, so that will require some closer evaluation.

bbende commented:
Thanks @exceptionfactory ! I noticed the CodeQL issues as well...

I did sanitize the Filename header value before passing it to any code that would use it to write a File. 

If that isn't enough then I suppose a fall back option would be to write the file names with the generated UUID and store some kind of mapping file along side the asset that contains the original filename. 

The problem with this is that the value of the parameter that references the asset is going to be the concatenated paths of the assets and it's going to contain file names that the user has no idea what they are. Example, upload PostgresDriver.jar and it generates id 123, then reference this asset in a parameter named db_driver, and the value of the parameter will be '/path/to/assets/123'. So some looking at the parameter has no idea what file that is.

exceptionfactory commented:
Thanks for the reply @bbende. Several classes mitigate the path traversal problem using Java Path objects with normalization and checking that the resolved child path starts with the selected parent path.

https://github.com/apache/nifi/blob/83d2078a4f878ba742f8adba2da382fb117ad6c0/nifi-system-tests/nifi-system-test-extensions-bundle/nifi-system-test-extensions/src/main/java/org/apache/nifi/flow/registry/FileSystemFlowRegistryClient.java#L434

It appears that the parent Parameter Context ID also needs to be checked. Instead of using the user-provided value, recommend looking up the Parameter Context and using the ID returned.

bbende commented:
@exceptionfactory thanks for the pointers! I made a couple of improvements and seems to have resolved the CodeQL issues. I think most of them were actually somewhat false positives because the sanitize method was returning the input variable if null or blank, but that was causing CodeQL to think the unmodified user input was being potentially used later on.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -323,6 +356,204 @@ public Response updateParameterContext(
+        // different request for each of the two phases.
+
+        final long startTime = System.currentTimeMillis();
+        final InputStream maxLengthInputStream = new MaxLengthInputStream(assetContents, (long) DataUnit.GB.toB(1));
```
For tracking purposes, it would be helpful to promote the maximum length value to a static value. Although this could be made configurable, it seems unlikely to change, and 1 GB seems like it should be large enough to handle anything of reasonable size.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -323,6 +356,204 @@ public Response updateParameterContext(
+            assetEntity = uploadRequestReplicator.upload(uploadRequest);
+        } else {
+            final String existingContextId = contextEntity.getId();
+            final String sanitizedAssetName = FileUtils.sanitizeFilename(assetName);
```
Instead of simply sanitizing the filename, should an input filename be rejected if it contains unexpected characters?

bbende replied:
That probably does make sense, will play around with that.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/dto/DtoFactory.java:
```diff
@@ -1509,6 +1512,26 @@ private ParameterProviderConfigurationEntity createParameterProviderConfiguratio
+       dto.setId(asset.getIdentifier());
+       dto.setName(asset.getName());
+       dto.setDigest(asset.getDigest().orElse(null));
+       dto.setMissingContent(!assetFile.exists());
```
What do you think about naming this property `contentNotFound` instead of `missingContent`?

bbende replied:
I don't have a strong preference, it seemed like we used "missing" for other stuff like components that are ghosted.

exceptionfactory commented on nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/client/nifi/ParamContextClient.java:
```diff
@@ -40,4 +43,10 @@ public interface ParamContextClient {
+
+    AssetsEntity getAssets(String contextId) throws NiFiClientException, IOException;
+
+    File getAssetContent(String contextId, String assetId, File outputDirectory) throws NiFiClientException, IOException;
```
What do you think about returning a `Path` instead of `File` for this method?

exceptionfactory reviewed (commented):
Great work on this new feature @bbende! The majority of the code changes look good, and the initial changes to address the CodeQL path concerns also look good. I plan to do some additional testing, but on initial review, I just noted a couple minor things.

bbende commented:
@exceptionfactory I pushed a commit that addresses the comments you had, and also another commit that adds a delete end-point to be able to clean up assets that were uploaded and maybe never referenced, let me know of anything else

markobean commented:
Is this new feature accessible only via the API and not in the GUI?

bbende commented:
@markobean at the moment yes, if someone picks up a UI JIRA to add it then it may be in the UI later on

exceptionfactory reviewed (approved):
Thanks for building out these new features @bbende, the latest version looks good! +1 merging
